koku/api/iam/customer_manager.py

```diff
@@ -0,0 +1,92 @@
+#
+# Copyright 2018 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+"""Management capabilities for Customer functionality."""
+
+import logging
+
+from django.core.exceptions import ValidationError
+
+from api.iam.models import Customer
+from api.iam.serializers import CustomerSerializer
+
+LOG = logging.getLogger(__name__)
+
+
+class CustomerManagerError(Exception):
+    """General Exception class for CustomerManager errors."""
+
+    def __init__(self, message):
+        """Set custom error message for CustomerManager errors."""
+        self.message = message
+
+
+class CustomerManagerDoesNotExist(Exception):
+    """CustomerManager could not find customer error."""
+
+    def __init__(self, message):
+        """Set custom error message for CustomerManager errors."""
+        self.message = message
+
+
+class CustomerManagerPermissionError(Exception):
+    """CustomerManager permission error."""
+
+    def __init__(self, message):
+        """Set custom error message for CustomerManager errors."""
+        self.message = message
+
+
+class CustomerManager:
+    """Customer Manager to manage operations related to customer tasks."""
+
+    def __init__(self, uuid):
+        """Establish provider manager database objects."""
+        self.model = None
+        try:
+            query = Customer.objects.all().filter(uuid=uuid)
+            self.model = query.get()
+        except Customer.DoesNotExist as e:
+            msg = 'Could not find customer with uuid: {}'.format(uuid)
+            raise CustomerManagerDoesNotExist(msg)
+        except ValidationError as e:
+            raise(CustomerManagerError(str(e)))
+
+    def get_model(self):
+        """Get the model object for the customer."""
+        return self.model
+
+    def get_name(self):
+        """Get the name of the customer."""
+        return self.model.name
+
+    def get_users_for_customer(self):
+        """List of users that belong to the customer."""
+        group = CustomerSerializer.get_authentication_group_for_customer(self.model)
+        users = CustomerSerializer.get_users_for_group(group)
+        return users
+
+    def remove_users(self, current_user):
+        """Remove users associated with this customer."""
```

Comment:
In retrospect it seems obvious now, but it took me a running through the tests for a few minutes for it to really sink in that this method always deletes all available users for the customer, and that it performs differently based on the user getting passed in. It may benefit from a follow up description in the doc string. 

Tangentially, we (the team) haven't discussed docstring style, with the exception of api and doc formatting in some docstrings. We could maybe benefit from landing on a particular docstring style? Plug: I find the Google docstring style to be quite readable.

Reply:
Yeah.  I definitely feel some strict style guidelines on docstrings is a great idea.  This is a good example to illustrate that we should look at a more uniform way to deal with permissions across koku.